lib/ocpp/v201/charge_point.cpp

```diff
@@ -3207,6 +3232,96 @@ void ChargePoint::handle_set_charging_profile_req(Call<SetChargingProfileRequest
+    const auto msg = call.msg;
+    ClearChargingProfileResponse response;
+    response.status = ClearChargingProfileStatusEnum::Unknown;
+
+    // K10.FR.06
+    if (msg.chargingProfileCriteria.has_value() and
+        msg.chargingProfileCriteria.value().chargingProfilePurpose.has_value() and
+        msg.chargingProfileCriteria.value().chargingProfilePurpose.value() ==
+            ChargingProfilePurposeEnum::ChargingStationExternalConstraints) {
+        response.statusInfo = StatusInfo();
+        response.statusInfo->reasonCode = "InvalidValue";
+        response.statusInfo->additionalInfo = "ChargingStationExternalConstraintsInClearChargingProfileRequest";
+        EVLOG_debug << "Rejecting SetChargingProfileRequest:\n reasonCode: " << response.statusInfo->reasonCode.get()
+                    << "\nadditionalInfo: " << response.statusInfo->additionalInfo->get();
+    } else {
+        response = this->smart_charging_handler->clear_profiles(msg);
+    }
+
+    ocpp::CallResult<ClearChargingProfileResponse> call_result(response, call.uniqueId);
+    this->send<ClearChargingProfileResponse>(call_result);
+}
+
+void ChargePoint::handle_get_charging_profiles_req(Call<GetChargingProfilesRequest> call) {
+    EVLOG_debug << "Received GetChargingProfilesRequest: " << call.msg << "\nwith messageId: " << call.uniqueId;
+    const auto msg = call.msg;
+    GetChargingProfilesResponse response;
+
+    const auto profiles_to_report = this->smart_charging_handler->get_reported_profiles(msg);
+
+    response.status =
+        profiles_to_report.empty() ? GetChargingProfileStatusEnum::NoProfiles : GetChargingProfileStatusEnum::Accepted;
+
+    ocpp::CallResult<GetChargingProfilesResponse> call_result(response, call.uniqueId);
+    this->send<GetChargingProfilesResponse>(call_result);
+
+    if (response.status == GetChargingProfileStatusEnum::NoProfiles) {
+        return;
+    }
+
+    // There are profiles to report.
+    // Prepare ReportChargingProfileRequest(s). The message defines the properties evseId and chargingLimitSource as
+    // required, so we can not report all profiles in a single ReportChargingProfilesRequest. We need to prepare a
+    // single ReportChargingProfilesRequest for each combination of evseId and chargingLimitSource
+    std::set<int32_t> evse_ids;                // will contain all evse_ids of the profiles
+    std::set<ChargingLimitSourceEnum> sources; // will contain all sources of the profiles
+
+    // fill evse_ids and sources sets
+    std::transform(profiles_to_report.begin(), profiles_to_report.end(), std::inserter(evse_ids, evse_ids.end()),
+                   [](const ReportedChargingProfile& profile) { return profile.evse_id; });
+    std::transform(profiles_to_report.begin(), profiles_to_report.end(), std::inserter(sources, sources.end()),
+                   [](const ReportedChargingProfile& profile) { return profile.source; });
```

Comment:
While I do appreciate a good algorithm, this could maybe be simpler with a small for loop like this:
```suggestion
    for(const auto& profile: profiles to report) {
        evse_ids.insert(profile.evse_id);
        sources.insert(profile.source);
    }
```